Fix lyrics Unicode corruption and escaped quotes in Genius plugin

grillofran opened the pull request:
## Problem
The lyrics plugin has two bugs that corrupt fetched lyrics:

1. **Unicode corruption**: Characters like `ò`, `è`, `à` are corrupted to `√≤`, `√®`, etc.
2. **Escaped quotes**: Quotes appear as `\"` instead of `"` in lyrics

## Root Causes

### Issue 1: MacRoman encoding misdetection
- **Location**: `RequestHandler.fetch_text()` line 220
- **Cause**: Setting `r.encoding = None` forces requests to use `apparent_encoding`
- **Problem**: For Genius.com (and others), requests incorrectly detects MacRoman instead of UTF-8
- **Result**: UTF-8 bytes `c3 b2` (ò) decoded as MacRoman produces "√≤" (U+221A U+2264)

### Issue 2: Incomplete JSON unescape
- **Location**: `Genius.scrape()` line 576
- **Cause**: The `remove_backslash` regex doesn't handle all escape patterns in JSON
- **Problem**: Genius embeds lyrics in JSON with patterns like `\\"` and `\\\\"` 
- **Result**: After BeautifulSoup processing, escaped quotes remain in final text

## Solution

### Fix 1: Trust server encoding, fallback to UTF-8
```python
# OLD: r.encoding = None
# NEW:
if not r.encoding:
    r.encoding = 'utf-8'
```
- Respects server's declared encoding (UTF-8 for Genius)
- Falls back to UTF-8 if no encoding specified (safer than apparent_encoding)
- Preserves original intent of handling misconfigured servers

### Fix 2: Iteratively clean escaped quotes
```python
while '\\"' in lyrics:
    lyrics = lyrics.replace('\\"', '"')
```
- Handles variable escape levels (`\"`, `\\\"`, `\\\\\"`)
- Minimal change - keeps original `remove_backslash` regex
- Applied after BeautifulSoup to avoid interfering with HTML parsing

## Testing

Tested with:
- Caparezza - "Argenti Vive" (Italian, many accented characters)
- WestsideGunn - "Heel Cena" (escaped quotes in lyrics)

Before:
```
mi si par√≤ davanti
\\"I got big moves\\"
```

After:
```
mi si parò davanti
"I got big moves"
```

## Impact
- Fixes lyrics for all languages with non-ASCII characters
- Fixes Genius lyrics with quotes
- No breaking changes - maintains backward compatibility
- Minimal code changes (14 lines total)

grillofran committed:
Fix lyrics Unicode corruption and escaped quotes in Genius plugin

## Problem
The lyrics plugin has two bugs that corrupt fetched lyrics:

1. **Unicode corruption**: Characters like `ò`, `è`, `à` are corrupted to `√≤`, `√®`, etc.
2. **Escaped quotes**: Quotes appear as `\"` instead of `"` in lyrics

## Root Causes

### Issue 1: MacRoman encoding misdetection
- **Location**: `RequestHandler.fetch_text()` line 220
- **Cause**: Setting `r.encoding = None` forces requests to use `apparent_encoding`
- **Problem**: For Genius.com (and others), requests incorrectly detects MacRoman instead of UTF-8
- **Result**: UTF-8 bytes `c3 b2` (ò) decoded as MacRoman produces "√≤" (U+221A U+2264)

### Issue 2: Incomplete JSON unescape
- **Location**: `Genius.scrape()` line 576
- **Cause**: The `remove_backslash` regex doesn't handle all escape patterns in JSON
- **Problem**: Genius embeds lyrics in JSON with patterns like `\\"` and `\\\\"` 
- **Result**: After BeautifulSoup processing, escaped quotes remain in final text

## Solution

### Fix 1: Trust server encoding, fallback to UTF-8
```python
# OLD: r.encoding = None
# NEW:
if not r.encoding:
    r.encoding = 'utf-8'
```
- Respects server's declared encoding (UTF-8 for Genius)
- Falls back to UTF-8 if no encoding specified (safer than apparent_encoding)
- Preserves original intent of handling misconfigured servers

### Fix 2: Iteratively clean escaped quotes
```python
while '\\"' in lyrics:
    lyrics = lyrics.replace('\\"', '"')
```
- Handles variable escape levels (`\"`, `\\\"`, `\\\\\"`)
- Minimal change - keeps original `remove_backslash` regex
- Applied after BeautifulSoup to avoid interfering with HTML parsing

## Testing

Tested with:
- Caparezza - "Argenti Vive" (Italian, many accented characters)
- WestsideGunn - "Heel Cena" (escaped quotes in lyrics)

Before:
```
mi si par√≤ davanti
\\"I got big moves\\"
```

After:
```
mi si parò davanti
"I got big moves"
```

## Impact
- Fixes lyrics for all languages with non-ASCII characters
- Fixes Genius lyrics with quotes
- No breaking changes - maintains backward compatibility
- Minimal code changes (14 lines total)

grillofran committed:
Update lyrics.py

## Key Improvements (addressing reviewer feedback):
- ✅ Uses regex instead of while loop (more efficient, one pass)
- ✅ No infinite loop risk
- ✅ Handles any number of backslashes before quotes
- ✅ Clear inline comments

grillofran committed:
Minor linting change

grillofran committed:
Finalized linting changes

  In the fetch_text method (line ~207):
  if not r.encoding:
      r.encoding = "utf-8"  # ← Double quotes!

  In the scrape method, make sure there are 2 blank lines before the next class:
          return None


  class Tekstowo(SearchBackend):  # ← Two blank lines above

  That should pass the formatting check! The repo follows PEP 8 style (double quotes, 2 blank lines between classes).

semohr commented on beetsplug/lyrics.py:
```diff
@@ -557,7 +561,11 @@ def search(self, artist: str, title: str) -> Iterable[SearchResult]:
             html_text = cls.remove_backslash(m[0]).replace(r"\n", "\n")
-            return cls.get_soup(html_text).get_text().strip()
+            lyrics = cls.get_soup(html_text).get_text().strip()
+            # Clean up any remaining escaped quotes (may need multiple passes)
```
Comments should be minimal and focus on the why rather than the how:
```
#  Strip HTML and normalize quotes that were left because of JSON...
return re.sub(r'\\+"', '"', lyrics)
```
It should explain why the escaped quotes exist without repeating how the regex works.

semohr reviewed (changes requested):
Thank you for the PR. Since this seems to be your first contribution to beets, you might want to have a look at our [developer guides](https://beets.readthedocs.io/en/stable/dev/index.html). Welcome aboard!

> Is the 66% patch coverage requirement strict for bug fixes? The changes are small edge-case improvements. Happy to add tests if required!

While it is not a requirement, for new code tests are always highly appreciated. Here for example showing that JSON encoded lyrics can be parsed as expected would be a good addition.  Same for the inverse case showing that "normal" html lyrics still work as expected would be good (I guess the test already exists, havent look at our testsuit 😁 )

----

More of a general question: how come this was never raised as an issue? Seems like it should be a common issue to me. Have you tried to find related open issues yet?

snejus commented on beetsplug/lyrics.py:
```diff
@@ -200,7 +200,11 @@ def get_text(
         self.debug("Fetching HTML from {}", url)
         r = self.get(url, **kwargs)
-        r.encoding = None
+        """Trust server's encoding,
```
@grillofran setting encoding to `None` is required for some websites under `Google` provider, such as **AZLyrics**. Your fix should apply to `Genius` backend only.

You could potentially add a new parameter `reset_encoding: bool=False` to `LyricsRequestHandler.get_text`, and adjust `Google.get_text` accordingly?

grillofran committed:
address pr feedback

grillofran committed:
Merge branch 'master' into patch-1

grillofran committed:
fix formatting and type issues

grillofran committed:
fix type signature

grillofran commented:
Thanks for the feedback! I've updated the PR to address all your points.

- Made the encoding fix Genius-only with a force_utf8 parameter (thanks @snejus)
- Simplified the encoding check to r.encoding = r.encoding or "utf-8" (thanks @semohr)
- Updated the comment to focus on why the escapes exist
- Added test cases for Caparezza (Italian accents) and Arctic Monkeys (quoted lyrics), as the other one was... too explicit

Funny enough.... I couldn't find any issues reporting this in the repo's issue tracker, which I thought was strange given how many people must use Genius. Also, this is my first contribution to beets so apologies if I missed anything. Still learning the ropes! Let me know if there's anything else that needs adjusting.

snejus commented on beetsplug/lyrics.py:
```diff
@@ -190,17 +190,23 @@ def format_url(url: str, params: JSONDict | None) -> str:
         r = self.get(url, **kwargs)
-        r.encoding = None
+        if force_utf8:
+            r.encoding = r.encoding or "utf-8"
```
This only handles `Genius` requirements: to clarify, for `Google`, we want:

```py
r.encoding = None
```

same it was before.

snejus replied:
Luckily, I managed to dig this URL out from my `ipython` history. Add this to `test/plugins/lyrics_pages.py`

```py
    LyricsPage.make(
        "https://www.azlyrics.com/lyrics/derivakat/bountyhunter.html",
        """
        [Romanized:]

        Living legend made of myths and made of stories
        But I'm present, find the mark and I'll take the glory
        Hit the dead-end, cornered like an animal, it just
        Takes one second, I'm nothing if not practical
        But I don't really care if you're right or you're wrong
        Bùguǎn nǐ de wùhuì or you're just a disgrace
        Get in the zone when I put on my song
        Bìrán de xiànshí when I'm up in your face
        Zhēn tài kě qíng kùn zài lóng lǐ de nǐ
        But you'll find in the end that I'll leave you erased
        (If they call?) I'm the one
        (Got a job?) Get it done
        All alone, you and me, I'll be leaving no trace

        No place where I won't go to
        No limits I won't break through
        No godless deed I won't do
        If they got problems, then they know to

        Call the bounty hunter
        Yeah, they got my number
        I'm the problem solver
        Leave you six feet under
        You won't see another
        Winter, spring or summer
        Wǒ huì ràng nǐ xiāoshī
        No way to recover (Hey!)

        Run it up
        Night or day (Hey!)
        Hunt you down
        Run away

        No mercy from this mercenary
        Take away your sanctuary
        Shìlì jiù xiàng 20/20
        Way I see you'll end up buried
        Don't deviate
        Zhǎodào nǐ so I terminate
        Wúlùn, wǒ huì make you pay
        I'll double tap you like bang, bang

        (Dead lilac) A ghost you only heard in rumor
        (Dead lilac) But only if you heard me sooner
        (Dead lilac) Losing your chances and losing your time
        (Dead lilac) shìjiè wèilái méi nǐ de cúnzài (Yeah)

        No place where I won't go to (I won't go to)
        No limits I won't break through (I won't break through)
        No godless deed I won't do (That I won't do)
        If they got problems, then they know to

        Call the bounty hunter (Hey)
        Yeah, they got my number
        I'm the problem solver (Hey)
        Leave you six feet under
        You won't see another (Hey)
        Winter, spring or summer
        Wǒ huì ràng nǐ xiāoshī (Hey)
        No way to recover (Hey!)

        Run it up (Run it up)
        Night or day (Hey!) (Day)
        Hunt you down (Hunt you down)
        Run away

        [English/Chinese:]

        Living legend made of myths and made of stories
        But I'm present, find the mark and I'll take the glory
        Hit the dead-end, cornered like an animal, it just
        Takes one second, I'm nothing if not practical
        But I don't really care if you're right or you're wrong
        不管你的误会 or you're just a disgrace
        Get in the zone when I put on my song
        必然的现实 when I'm up in your face
        真太可情 困在笼里的你
        But you'll find in the end that I'll leave you erased
        (If they call?) I'm the one
        (Got a job?) Get it done
        All alone, you and me, I'll be leaving no trace

        No place where I won't go to
        No limits I won't break through
        No godless deed I won't do
        If they got problems, then they know to

        Call the bounty hunter
        Yeah, they got my number
        I'm the problem solver
        Leave you six feet under
        You won't see another
        Winter, spring or summer
        我会让你消失
        No way to recover (Hey!)

        Run it up
        Night or day (Hey!)
        Hunt you down
        Run away

        No mercy from this mercenary
        Take away your sanctuary
        视力就像 20/20
        Way I see you'll end up buried
        Don't deviate
        找到你 so I terminate
        无论, 我会 make you pay
        I'll double tap you like bang, bang

        (Dead lilac) A ghost you only heard in rumor
        (Dead lilac) But only if you heard me sooner
        (Dead lilac) Losing your chances and losing your time
        (Dead lilac) 世界未来没你的存在 (Yeah)

        No place where I won't go to (I won't go to)
        No limits I won't break through (I won't break through)
        No godless deed I won't do (That I won't do)
        If they got problems, then they know to

        Call the bounty hunter (Hey)
        Yeah, they got my number
        I'm the problem solver (Hey)
        Leave you six feet under
        You won't see another (Hey)
        Winter, spring or summer
        我会让你消失 (Hey)
        No way to recover (Hey!)

        Run it up (Run it up)
        Night or day (Hey!) (Day)
        Hunt you down (Hunt you down)
        Run away
        """,
        artist="Derivakat",
        track_title="Bounty Hunter",
        url_title="Derivakat - Bounty Hunter Lyrics | AZLyrics.com",
        marks=[xfail_on_ci("AZLyrics is blocked by Cloudflare")],
    ),
```

and run

```bash
$ LYRICS_UPDATED=true poe test test/plugins/test_lyrics.py -k azlyrics -vv
```

[![asciicast](https://asciinema.org/a/2Qqp0Nd1imH3eemVc6GTivznc.svg)](https://asciinema.org/a/2Qqp0Nd1imH3eemVc6GTivznc)

grillofran committed:
add azlyrics test case

grillofran committed:
Merge branch 'master' into patch-1

grillofran committed:
set encoding to None for non-genius backends

grillofran commented:
Hey folks (happy holidays)

Ah, got it. I think I had misunderstood. I added the `else: r.encoding = None` to restore the original behavior for Google/AZLyrics. Should fix the Chinese character issue in the test. Setting `r.encoding = None` explicitly should fix that while keeping the UTF-8 fix for Genius only. Thoughts?

Thanks!
